File: Rules.php
<?php
class Rules
{
	/* Connection to MySQL database */
	private $connection;

	/* Private properties */
	private $_description; /* Exposed via getDescription() */
	private $_preprocess;
	private $_postprocess;
	private $_query;
	private $_message;
	private $_output; /* Exposed via getOutput() */
	private $_active; /* Exposed via getActive() */

	private function reset()
	{
		$this->_description = NULL;
		$this->_preprocess = NULL;
		$this->_postprocess = NULL;
		$this->_query = NULL;
		$this->_message = NULL;
		$this->_output = NULL;
		$this->_active = NULL;
		$this->_shellcmd = NULL;
		$this->_runshell = NULL;
	}

	private function connect( $host, $username, $password, $database )
	{
		$this->connection = new mysqli( $host, $username, $password, $database );
	}


	private function disconnect()
	{
		$this->connection->close();
	}


	/* Get rule data */
	private function get_rule_data( $ruleid )
	{
		if ( $results = $this->connection->query( "select rule_desc
			                                       ,      rule_query
			                                       ,      rule_message
			                                       ,      rule_active
			                                       ,      rule_preproc
			                                       ,      rule_postproc
			                                       ,      rule_shellcmd
			                                       ,      rule_run_shell
			                                       from   rules
			                                       where  rule_id = ${ruleid}" ) ) {
			$row = mysqli_fetch_array( $results );
			$this->_description = $row[0];
			$this->_query = $row[1];
			$this->_message = $row[2];
			$this->_active = $row[3];
			$this->_preprocess = $row[4];
			$this->_postprocess = $row[5];
			$this->_shellcmd = $row[6];
			$this->_runshell = $row[7];
		}
	}


	public function run_rule( $ruleid )
	{
		/* Reset */
		$this->reset();

		/* Get rule data */
		$this->get_rule_data( $ruleid );

		if ( $this->_active == 'Y' ) {
			/* Clear any previous output */
			$this->_output = NULL;

			/* Load last used data into variable */
			$this->connection->query( "select rule_last_used into @last_used from rules where rule_id = ${ruleid}" );

			/* Execute pre process */
			if ( strlen( $this->_preprocess ) > 0 ) {
				$this->connection->query( $this->_preprocess );
			}

			/* Execute the actual rule */
			$rule_result = $this->connection->query( $this->_query );

			/* Format the message */
			if ( mysqli_num_rows( $rule_result ) == 1 ) {
				$this->_output = vsprintf( $this->_message, mysqli_fetch_array( $rule_result, MYSQLI_NUM ) );
			}

			/* Execute post process */
			if ( strlen( $this->_postprocess ) > 0 ) {
				$this->connection->query( $this->_postprocess );
			}

			/* Update last usage timestamp for rule */
			$this->connection->query( "update rules set rule_last_used=now() where rule_id = ${ruleid}" );
			$this->connection->commit();

			/* Remove variable */
			$this->connection->query( "set @last_used = null;" );

			/* Result result message */
			return ( strlen( $this->_output ) > 0 );
		} 

		/* Return failed indicator */
		return false;
	}


	/* Return the value of the private property _ */
	public function getDescription()
	{
		return $this->_description;
	}


	/* Return the value of the private property _output */
	public function getOutput()
	{
		return $this->_output;
	}


	/* Execute the shell command specified in _shellcmd */
	public function runShellCmd()
	{
		if ( strlen( $this->_shellcmd ) > 0 ) {
			switch ( $this->_runshell ) {
				case 'never':
					// Nothing to do
					break;

				case 'always':
					shell_exec( $this->_shellcmd );
					break;

				case 'results':
					if ( strlen( $this->_output ) > 0 ) {
						shell_exec( $this->_shellcmd );					
					}
					break;
			}
		}
	}


	/* Return the value of the private property _active */
	public function getActive()
	{
		return $this->_active;
	}


	/* Default Constructor */
	public function __construct( $host, $username, $password, $database )
	{
		$this->connect( $host, $username, $password, $database );
	}


	/* Default Destructor */
   function __destruct() {
   		$this->disconnect();
   }
}
?>

File: Rules.py
# -*- coding: utf-8 -*- 

import MySQLdb as sql
import os

class Rules:
	__description = None
	__preprocess = None
	__postprocess = None
	__query = None
	__message = None
	__output = None
	__active = None
	__shellcmd = None
	__runshell = None


	def reset( self ):
		self.__description = None
		self.__preprocess = None
		self.__postprocess = None
		self.__query = None
		self.__message = None
		self.__output = None
		self.__active = None
		self.__shellcmd = None
		self.__runshell = None


	def connect( self, host, username, password, database ):
		self.connection = sql.connect( host, username, password, database )
		self.cursor = self.connection.cursor()


	def disconnect( self ):
		self.connection.close()


	def get_rule_data( self, ruleid ):
		self.cursor.execute( "select rule_desc \
                              ,      rule_query \
                              ,      rule_message \
                              ,      rule_active \
                              ,      rule_preproc \
                              ,      rule_postproc \
                              ,      rule_shellcmd \
                              ,      rule_run_shell \
                              from   rules \
                              where  rule_id = {0}".format( ruleid ) )

		row = self.cursor.fetchone()

		self.__description = row[0]
		self.__query = row[1]
		self.__message = row[2]
		self.__active = row[3]
		self.__preprocess = row[4]
		self.__postprocess = row[5]
		self.__shellcmd = row[6]
		self.__runshell = row[7]


	def run_rule( self, ruleid ):
		self.reset()
		self.get_rule_data( ruleid )

		if ( self.__active == 'Y' ):
			# Clear any previous output
			self.__output = None

			# Load last used data into variable
			self.cursor.execute( "select rule_last_used into @last_used from rules where rule_id = {0}".format( ruleid ) )

			# Execute pre process
			if ( self.__preprocess is not None and len( self.__preprocess ) > 0 ):
				self.cursor.execute( self.__preprocess )

			# Execute the actual rule
			rule_result = self.cursor.execute( self.__query )

			# Format the message
			if ( self.cursor.rowcount == 1 ):
				self.__output = self.__message % ( self.cursor.fetchone() )

			# Execute post process
			if ( self.__postprocess is not None and len( self.__postprocess ) > 0 ):
				self.cursor.execute( self.__postprocess )

			# Update last usage timestamp for rule
			self.cursor.execute( "update rules set rule_last_used=now() where rule_id = {0}".format( ruleid ) )
			self.connection.commit()

			# Remove variable
			self.cursor.execute( "set @last_used = null;" )

			# Result result message
			return ( self.__output is not None and len( self.__output ) > 0 )

		# Return failed indicator
		return false


	def getDescription( self ):
		return self.__description


	def getOutput( self ):
		return self.__output


	def runShellCmd( self ):
		if ( self.__shellcmd is not None and len( self.__shellcmd ) > 0 ):
			if ( self.__runshell == 'never' ):
				# Nothing to do
				pass
			elif ( self.__runshell == 'always' ):
				os.system( self.__shellcmd )
			elif ( self.__runshell == 'results' ):
				if ( self.__output is not None and len( self.__output ) > 0 ):
					os.system( self.__shellcmd )


	def getActive( self ):
		return self.__active


	def __init__( self, host, username, password, database ):
		self.connect( host, username, password, database )


	def __del__( self ):
		self.disconnect()
